ssl/ech.c

```diff
@@ -0,0 +1,506 @@
+        || EVP_DigestInit_ex(mdctx, md, NULL) <= 0
+        || EVP_DigestUpdate(mdctx, buf, blen) <= 0
+        || EVP_DigestFinal_ex(mdctx, hashval, &hashlen) <= 0) {
+        if (md != NULL)
+            EVP_MD_free(md);
+        if (mdctx != NULL)
+            EVP_MD_CTX_free(mdctx);
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    EVP_MD_free(md);
+    EVP_MD_CTX_free(mdctx);
+    if (ah_encode(ah_hash, ah_len, hashval, hashlen) != 1)
+        return 0;
+    return 1;
+}
+
+/* 
+ * API calls build around OSSL_ECHSSTORE 
+ */
+
+OSSL_ECHSTORE *OSSL_ECHSTORE_init(OSSL_LIB_CTX *libctx, const char *propq)
+{
+    OSSL_ECHSTORE *es = NULL;
+
+    es = OPENSSL_zalloc(sizeof(OSSL_ECHSTORE));
+    if (es == NULL) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    es->libctx = libctx;
+    es->propq = propq;
+    return es;
+}
+
+static void OSSL_ECHEXT_free(OSSL_ECHEXT *e)
+{
+    OPENSSL_free(e->val);
+    OPENSSL_free(e);
+    return;
+}
+
+static void OSSL_ECHSTORE_entry_free(OSSL_ECHSTORE_entry *ee)
+{
+    OPENSSL_free(ee->public_name);
+    OPENSSL_free(ee->pub);
+    OPENSSL_free(ee->pemfname);
+    EVP_PKEY_free(ee->keyshare);
+    OPENSSL_free(ee->encoded);
+    OPENSSL_free(ee->suites);
+    sk_OSSL_ECHEXT_pop_free(ee->exts, OSSL_ECHEXT_free);
+    OPENSSL_free(ee);
+    return;
+}
+
+void OSSL_ECHSTORE_free(OSSL_ECHSTORE *es)
+{
+    sk_OSSL_ECHSTORE_entry_pop_free(es->entries, OSSL_ECHSTORE_entry_free);
+    OPENSSL_free(es);
+    return;
+}
+
+int OSSL_ECHSTORE_new_config(OSSL_ECHSTORE *es,
+                             uint16_t echversion, uint8_t max_name_length,
+                             const char *public_name, OSSL_HPKE_SUITE suite)
+{
+    size_t pnlen = 0;
+    size_t publen = OSSL_ECH_CRYPTO_VAR_SIZE;
+    unsigned char pub[OSSL_ECH_CRYPTO_VAR_SIZE];
+    int rv = 0;
+    unsigned char *bp = NULL;
+    size_t bblen = 0;
+    EVP_PKEY *privp = NULL;
+    uint8_t config_id = 0;
+    WPACKET epkt;
+    BUF_MEM *epkt_mem = NULL;
+    OSSL_ECHSTORE_entry *ee = NULL;
+    char pembuf[2 * EVP_MAX_MD_SIZE + 1];
+    size_t pembuflen = 2 * EVP_MAX_MD_SIZE + 1;
+
+    /* basic checks */
+    if (es == NULL) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    pnlen = (public_name == NULL ? 0 : strlen(public_name));
+    if (pnlen > OSSL_ECH_MAX_PUBLICNAME
+        || max_name_length > OSSL_ECH_MAX_MAXNAMELEN) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+    /* this used have more versions and will again in future */
+    switch (echversion) {
+    case OSSL_ECH_RFCXXXX_VERSION:
+        break;
+    default:
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+
+    /* so WPAKCET_cleanup() won't go wrong */
+    memset(&epkt, 0, sizeof(epkt));
+    /* random config_id */
+    if (RAND_bytes((unsigned char *)&config_id, 1) <= 0) {
```

Comment:
Use `RAND_bytes_ex` and pass the correct libctx/propq

Reply:
ack